### PR TITLE
gsl_ran_flat() patch

### PR DESCRIPTION
**Pull request by chriscrsmith**
Pull request to solve gsl_ran_flat() bug #1997

### CODE REVIEW

File: lib/tests/test_core.c [hunk 1]
```diff
@@ -130,6 +131,45 @@ test_tskit_version(void)
...
+    gsl_rng_set(rng, 18112);
+    FILE *rng_state;
+    rng_state = fopen("../tests/test_core_rngstate0.bin", "r");
+    gsl_rng_fread(rng_state, rng);
+    fclose(rng_state);
+    left = 85189995.000000;
+    right = 85190021.000000;
+    x = msp_gsl_ran_flat(rng, left, right);
+    CU_ASSERT(left <= x);
+    CU_ASSERT(right > x);
```
Comment (andrewkern) [inline lib/tests/test_core.c:151]: it might be good to run this to _create_ the error too. e.g.,
```suggestion
    x = gsl_ran_flat(rng, left, right);
    CU_ASSERT(left <= x);
    CU_ASSERT(!(right > x));
```

**Review by molpopgen — COMMENTED** [review]
I think that the method used to trigger the bug is not safe. The binary versions of the random number generator state are not portable across platforms. So if they work in the CI testing, that is luck and not design.  I have an idea of how to trigger it. I hope to get a chance to test the idea tomorrow.

**Comment by molpopgen** [comment]
Quoted from the GSL docs:

```
This function w[r](https://www.gnu.org/software/gsl/doc/html/rng.html#c.gsl_rng_fwrite)ites the random number state of the random number generator r to the [stream](https://www.gnu.org/software/gsl/doc/html/rng.html#c.gsl_rng_fwrite) stream in binary format. The return value is 0 for success and GSL_EFAILED if there was a problem writing to the file. Since the data is written in the native binary format it may not be portable between different architectures.
```

**Comment by molpopgen** [comment]
Following my comments in #1997, this triggers the bug for any RNG seed.  Note the requirement for C99, which is okay since msprime is compiled with C99:

```c
// gcc  -std=c99 -o trigger trigger.c -lgsl -lgslcblas -lm  

#include <stdio.h>
#include <math.h>
#include <assert.h>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>

int main(int argc, char ** argv)
{
    const double hi = 85190021.000000;

    // Get the next double different from hi
    // in the direction towards 0.0
    const double next_to_zero = nexttoward(hi, 0.0);
    double rv;
    gsl_rng * r = gsl_rng_alloc(gsl_rng_default);
    gsl_rng_set(r, 12345);

    // This will  print 0.0 for the last value (on most systems at least)...
    fprintf(stdout,"%lf %lf %lf\n", hi, next_to_zero, hi - next_to_zero);

    // ... but it is not 0.0!!
    assert(next_to_zero < hi);
    assert(next_to_zero != hi);
    rv = gsl_ran_flat(r, next_to_zero, hi);

    assert(rv > next_to_zero);
    assert(rv == hi);
}
```

**Comment by molpopgen** [comment]
It will be interesting to know if the rejection sampling approach leads to an infinite loop or not for the low end high values used in my example to trigger the bug.  I'll hazard a guess that it will, for at least some  floating point values with large integer components.

**Comment by molpopgen** [comment]
> It will be interesting to know if the rejection sampling approach leads to an infinite loop or not for the low end high values used in my example to trigger the bug.  I'll hazard a guess that it will, for at least some  floating point values with large integer components.

Actually it won't be an infinite loop now that I think about it more. But it could be very long, requiring waiting until the random number generator returns zero exactly.

**Comment by chriscrsmith** [comment]
@molpopgen is this how you envisioned the test? This loops just a handful of times (1-10)

`#include <stdio.h>
#include <math.h>
#include <assert.h>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>

int main(int argc, char ** argv)
{
  const double hi = 85190021.000000;
  const double next_to_zero = nexttoward(hi, 0.0);
  double rv;
  gsl_rng * r = gsl_rng_alloc(gsl_rng_default);
  gsl_rng_set(r, 12345);

  do {
    rv = gsl_ran_flat(r, next_to_zero, hi);
    fprintf(stdout,"%lf %lf %lf\n", next_to_zero, hi, rv);
  } while (rv >= hi && next_to_zero != hi);

  assert(rv >= next_to_zero);
  assert(rv < hi);
}`

File: lib/util.c [hunk 2]
```diff
@@ -712,3 +713,24 @@ fast_search_free(fast_search_t *self)
...
+ *   This function includes a while loop to ensure x<hi.
+ *   https://github.com/tskit-dev/msprime/issues/1997
+ */
+double
+msp_gsl_ran_flat(gsl_rng *rng, double lo, double hi)
+{
+    double x;
+
+    if (lo > hi) {
+        double temp = lo;
```
Comment (jeromekelleher) [inline lib/util.c:728]: Style nitpick: we do all the declarations at the start of the function, so here it'd be ``double x, temp;``.

No particularly good reason for it anymore tbh, but best to stick with the established style at this point.

File: lib/util.c [hunk 3]
```diff
@@ -712,3 +713,24 @@ fast_search_free(fast_search_t *self)
...
+        lo = hi;
```
Comment (jeromekelleher) [inline lib/util.c:729]: Is this the behaviour of gsl_ran_flag when lo > hi?

I'd be tempted to just put in a ``tsk_bug_assert(lo < hi);`` here. That's the assumption we're making in all the usages from msprime.

That might simplify things a bit too?

**Review by jeromekelleher — COMMENTED** [review]
Looks great, thanks @chriscrsmith! Some minor comments.

Mainly: can we ``tsk_bug_assert`` out the awkward corner cases?

**Comment by molpopgen** [comment]
> @molpopgen is this how you envisioned the test? This loops just a handful of times (1-10)
> 
> ```
> #include <stdio.h>
> #include <math.h>
> #include <assert.h>
> #include <gsl/gsl_rng.h>
> #include <gsl/gsl_randist.h>
> 
> int main(int argc, char ** argv)
> {
>   const double hi = 85190021.000000;
>   const double next_to_zero = nexttoward(hi, 0.0);
>   double rv;
>   gsl_rng * r = gsl_rng_alloc(gsl_rng_default);
>   gsl_rng_set(r, 12345);
> 
>   do {
>     rv = gsl_ran_flat(r, next_to_zero, hi);
>     fprintf(stdout,"%lf %lf %lf\n", next_to_zero, hi, rv);
>   } while (rv >= hi && next_to_zero != hi);
> 
>   assert(rv >= next_to_zero);
>   assert(rv < hi);
> }
> ```

Yeah this looks right. That's good that it doesn't loop many times. I think that makes sense--there must be many  small values close to zero or one that are smaller than the size of the range used here.  That's a good thing.

File: lib/util.c (hunk 3 above)
Comment (chriscrsmith) [reply lib/util.c:729]: I do think this is what's going on with gsl_ran_flat when lo>hi. I think the `tsk_bug_assert(lo < hi);` makes sense

**Comment by chriscrsmith** [comment]
Hey @jeromekelleher and @molpopgen I incorporated your changes and squashed my commits.

**Comment by molpopgen** [comment]
I'm a bit concerned about checking for `lo == hi`.  I think we may want something more like the function below?  @jeromekelleher?

```c
double safe_gsl_ran_flat(const gsl_rng * r, double lo, double hi) {
    double rv;
    if (hi < lo + nextafter(0.0,1.0)) {
        return lo;
    }
    rv = gsl_ran_flat(r,lo,hi);
    while(!(hi > rv)) {
        rv = gsl_ran_flat(r,lo,hi);
    }
    return rv;
}
```

File: lib/tests/test_core.c [hunk 4]
```diff
@@ -130,6 +130,49 @@ test_tskit_version(void)
...
+        CU_ASSERT(left <= x);
+        CU_ASSERT(right > x);
+    }
+
+    left = 2.0;
+    right = 2.0;
+    x = msp_gsl_ran_flat(rng, left, right);
+    CU_ASSERT(left == x && right == x);
+
+    left = 3.0;
```
Comment (jeromekelleher) [inline lib/tests/test_core.c:164]: Ooh, clever! I think this might end up being a bit brittle though, so maybe not worth the complication? I wouldn't bother personally, we're not going to remove the assert from the msp_gsl_ran_flat function and that's all this is testing essentially.

File: lib/util.c [hunk 5]
```diff
@@ -712,3 +713,24 @@ fast_search_free(fast_search_t *self)
...
+/*   The gsl_ran_flat() function is supposed to output lo<=x<hi, but
+ *   sometimes (with probability <1e-9) we find x=hi.
+ *   This function includes a while loop to ensure x<hi.
+ *   https://github.com/tskit-dev/msprime/issues/1997
+ */
+double
+msp_gsl_ran_flat(gsl_rng *rng, double lo, double hi)
+{
+    double x;
+
```
Comment (jeromekelleher) [inline lib/util.c:726]: I guess a comment here saying "The logic here requires that lo <= hi, but gsl_ran_flat produces output in this case. Keep this assert to make sure that msprime's usage has the required properties. "

**Review by jeromekelleher — APPROVED** [review]
Looks great, thanks @chriscrsmith!

I don't think there's any issue with using "==" @molpopgen, I think this is good. 

In general, I don't see much point in doing things like ``!(hi > rv)`` instead of ``hi <= rv`` --- IEEE comparison semantics are well defined, aren't they? We are assuming IEEE in many places already.

**Comment by molpopgen** [comment]
> Alright @jeromekelleher and @molpopgen, here is an updated version based on your suggestions.
> 
> 

I'm happy with this pull request. Unless Jerome has any comments, I'm happy with a squash and merge at this point.

**Comment by jeromekelleher** [comment]
@mergifyio rebase

**Comment by andrewkern** [comment]
congrats on your first PR @chriscrsmith 🎉
